modules/pallets/messaging-fees/src/impls.rs

```diff
@@ -0,0 +1,293 @@
+
+			if let Some(per_byte_fee) = Self::get_per_byte_fee(&state_machine) {
+				let dollar_cost = per_byte_fee.saturating_mul(U256::from(bytes_processed));
+				let dollar_cost: u128 =
+					dollar_cost.try_into().map_err(|_| Error::<T>::CalculationOverflow)?;
+
+				let bridge_usd_price = T::PriceOracle::get_bridge_price()
+					.map_err(|_| Error::<T>::ErrorInPriceConversion)?;
+
+				let bridge_usd_price: u128 =
+					bridge_usd_price.try_into().map_err(|_| Error::<T>::CalculationOverflow)?;
+
+				let base_reward_in_token_u256 = bridge_usd_price
+					.checked_mul(dollar_cost.into())
+					.ok_or(Error::<T>::CalculationOverflow)?
+					.checked_div(DECIMALS.into())
+					.ok_or(Error::<T>::CalculationOverflow)?;
+
+				let base_reward_in_token: T::Balance = base_reward_in_token_u256
+					.try_into()
+					.map_err(|_| Error::<T>::CalculationOverflow)?;
+
+				let target_message_size = T::TargetMessageSize::get();
+
+				if current_total_bytes <= target_message_size {
+					let reward_amount = Self::calculate_reward(
+						current_total_bytes,
+						target_message_size,
+						base_reward_in_token.into(),
+					)?;
+
+					T::Currency::transfer(
+						&T::TreasuryAccount::get().into_account_truncating(),
+						&relayer,
+						reward_amount.saturated_into(),
+						Preservation::Expendable,
+					)
+					.map_err(|_| Error::<T>::RewardTransferFailed)?;
+					Self::deposit_event(Event::FeeRewarded {
+						relayer: relayer.clone(),
+						amount: reward_amount.saturated_into(),
+					});
+				} else {
+					T::Currency::transfer(
+						&relayer,
+						&T::TreasuryAccount::get().into_account_truncating(),
+						base_reward_in_token.saturated_into(),
+						Preservation::Expendable,
+					)
+					.map_err(|_| Error::<T>::RewardTransferFailed)?;
+
+					Self::deposit_event(Event::FeePaid {
+						relayer: relayer.clone(),
+						amount: base_reward_in_token_u256.into(),
+					});
+				}
+			}
+		}
+
+		Ok(())
+	}
+
+	/// A curve for calculating reward
+	/// Reward=BaseReward×((TargetSize−TotalBytes)/TargetSize)^2
+	pub fn calculate_reward(
+		total_bytes: u32,
+		target_size: u32,
+		base_reward: u128,
+	) -> Result<u128, Error<T>> {
+		if total_bytes >= target_size || target_size.is_zero() {
+			return Ok(u128::zero());
+		}
+
+		let total_bytes_u128 = total_bytes as u128;
+		let target_size_u128 = target_size as u128;
+
+		let decay_numerator = target_size_u128.saturating_sub(total_bytes_u128);
+
+		let decay_numerator_sq = decay_numerator
+			.checked_mul(decay_numerator)
+			.ok_or(Error::<T>::CalculationOverflow)?;
+
+		let target_size_sq = target_size_u128
+			.checked_mul(target_size_u128)
+			.ok_or(Error::<T>::CalculationOverflow)?;
+
+		let final_reward_numerator = base_reward
+			.checked_mul(decay_numerator_sq)
+			.ok_or(Error::<T>::CalculationOverflow)?;
+
+		Ok(final_reward_numerator.saturating_div(target_size_sq))
+	}
+	pub fn note_request_fee(commitment: H256, fee: u128) {
```

Comment:
Shouldn't this be implemented in the trait?

Reply:
the trait will be implemented in the runtime, which then calls this function. you can take a look at how i did it in the test runtime [here](https://github.com/polytope-labs/hyperbridge/blob/dami/messaging-incentives/modules/pallets/testsuite/src/runtime.rs#L203) which is then being used [here](https://github.com/polytope-labs/hyperbridge/blob/dami/messaging-incentives/modules/pallets/testsuite/src/runtime.rs#L222)

Reply:
Implement it directly on the pallet here, then you can use the pallet  in the runtime